src/setupWorker/setupWorker.ts

```diff
@@ -1,238 +1,226 @@
-let listeners: Listener[] = []
-
 /**
- * Creates a new mock Service Worker registration
- * with the given request handlers.
- * @param {RequestHandler[]} requestHandlers List of request handlers
- * @see {@link https://mswjs.io/docs/api/setup-worker `setupWorker`}
+ * Concrete class to implement the SetupApi for the browser environment, uses the ServerWorker setup.
  */
-export function setupWorker(
-  ...requestHandlers: RequestHandler[]
-): SetupWorkerApi {
-  requestHandlers.forEach((handler) => {
-    if (Array.isArray(handler))
+export class SetupWorkerApi extends SetupApi<WorkerLifecycleEventsMap> {
+  private context: SetupWorkerInternalContext
+  private startHandler: any
+  private stopHandler: any
+  private listeners: Listener[] = []
+
+  constructor(handlers: RequestHandler[]) {
+    super([], 'worker-setup', handlers)
+
+    if (isNodeProcess()) {
       throw new Error(
         devUtils.formatMessage(
-          'Failed to call "setupWorker" given an Array of request handlers (setupWorker([a, b])), expected to receive each handler individually: setupWorker(a, b).',
+          'Failed to execute `setupWorker` in a non-browser environment. Consider using `setupServer` for Node.js environment instead.',
         ),
       )
-  })
-
-  // Error when attempting to run this function in a Node.js environment.
-  if (isNodeProcess()) {
-    throw new Error(
-      devUtils.formatMessage(
-        'Failed to execute `setupWorker` in a non-browser environment. Consider using `setupServer` for Node.js environment instead.',
-      ),
-    )
-  }
+    }
 
-  const emitter = new StrictEventEmitter<WorkerLifecycleEventsMap>()
-  const publicEmitter = new StrictEventEmitter<WorkerLifecycleEventsMap>()
-  pipeEvents(emitter, publicEmitter)
-
-  const context: SetupWorkerInternalContext = {
-    // Mocking is not considered enabled until the worker
-    // signals back the successful activation event.
-    isMockingEnabled: false,
-    startOptions: undefined,
-    worker: null,
-    registration: null,
-    requestHandlers: [...requestHandlers],
-    emitter,
-    workerChannel: {
-      on(eventType, callback) {
-        context.events.addListener(
-          navigator.serviceWorker,
-          'message',
-          (event: MessageEvent) => {
-            // Avoid messages broadcasted from unrelated workers.
-            if (event.source !== context.worker) {
-              return
-            }
+    this.context = {} as SetupWorkerInternalContext
 
-            const message = event.data as ServiceWorkerMessage<
-              typeof eventType,
-              any
-            >
+    this.initializeWorkerContext()
+  }
 
-            if (!message) {
-              return
-            }
+  public initializeWorkerContext(): void {
+    this.context = {
+      // Mocking is not considered enabled until the worker
+      // signals back the successful activation event.
+      isMockingEnabled: false,
```

Comment:
This portion of the code is nearly untouched, I moved the full logic to a method for this new class. You will see that I had to recreate most of the base methods for the worker class because the logic for the worker was around the `this.context` object.

Reply:
No worries. I've removed some of the base class methods from `SetupWorkerApi` and made `this.context.requestHandlers` a getter, so it always returns the latest `this.currentHandlers` value inherited from the base class. I wish simple `this.X` reference would work but I assume since we're nesting it in another object (`this.context)` we lose the object magic. I'm fine with the getters.